src/ingest-pipeline/airflow/dags/catch_pipeline_errors.py

```diff
@@ -0,0 +1,163 @@
+import json
+from pprint import pprint
+from requests.exceptions import HTTPError
+from requests import codes
+import traceback
+
+from airflow.providers.http.hooks.http import HttpHook
+from airflow.utils.email import send_email
+
+from utils import get_auth_tok
+
+"""
+TODO
+- this requires crypt_auth_tok to be an op_kwarg, is that reasonable?
+- fix DRY issues (e.g. get_submission_context?, set_status)
+- type hinting
+"""
+
+
+class FailureCallbackException(Exception):
+    pass
+
+
+class FailureCallback:
+    """
+    List should be overwritten by each subclass with appropriate values.
+    """
+
+    external_exceptions = []
+    # TODO: Switch to curator email(s)
+    internal_email_recipients = ["gesina@psc.edu"]
+
+    def __init__(self, context, execute_methods=True):
+        self.context = context
+        self.uuid = self.context.get("task_instance").xcom_pull(key="uuid")
+        self.crypt_auth_tok = self.context.get("crypt_auth_tok")
+        self.dag_run = self.context.get("dag_run")
+        self.task = self.context.get("task")
+        self.exception = self.context.get("exception")
+        self.exception_name = type(self.exception).__name__
+
+        if execute_methods:
```

Comment:
This is fine for a first iteration as we've discussed. We will want to think through how we can make the functions that are called (set_status, send_failure_email) a bit more dynamic though as we add additional "reporting" methods. IE, updating Asana, Slack etc.

Reply:
Agreed, but I will do some thinking!

Reply:
Made an update here, but I'm not really sure if I'm moving in the right direction.